Libraries/MLXLMCommon/Evaluate.swift

```diff
@@ -0,0 +1,556 @@
+// Copyright © 2024 Apple Inc.
+
+import Foundation
+import MLX
+import MLXRandom
+import Tokenizers
+
+/// A `LogitSampler` is responsible for sampling `logits` produced by
+/// a ``LanguageModel`` to produce a token.
+///
+/// See also: ``LogitProcessor``
+public protocol LogitSampler: Sendable {
+
+    /// Given `logits` produce a new `MLXArray` with the token.
+    func sample(logits: MLXArray) -> MLXArray
+}
+
+/// A `LogitProcessor` is an optional visitor of `logits`.
+///
+/// The ``LogitProcessor`` is called with the input (prompt) before generating tokens:
+///
+/// ```swift
+/// processor?.prompt(input.text.tokens)
+/// ```
+///
+/// Then for each token generated it has a chance to adjust the logits:
+///
+/// ```swift
+/// logits = processor?.process(logits: logits) ?? logits
+/// let y = sampler.sample(logits: logits)
+/// processor?.didSample(token: y)
+/// ```
+///
+/// See also: ``LogitSampler``
+public protocol LogitProcessor: Sendable {
```

Comment:
These two protocols are added to describe the logit handling that happens.  Before this had been factored in to a struct  `GenerateParameters` but if you wanted to change the behavior there, good luck.  Now `GenerateParameters` has the same parameters (below) but it produces instances of these two protocols.  Callers could introduce new variants as desired or use `GenerateParameters` as they did before.

```swift
public struct GenerateParameters: Sendable {

    /// Step size for processing the prompt
    public var prefillStepSize = 512

    /// sampling temperature
    public var temperature: Float = 0.6

    /// top p sampling
    public var topP: Float = 1.0

    /// penalty factor for repeating tokens
    public var repetitionPenalty: Float?

    /// number of tokens to consider for repetition penalty
    public var repetitionContextSize: Int = 20
```